### PR TITLE
Add ambient typings for adapters of the smart home system ioBroker

### PR DESCRIPTION
**Pull request by AlCalzone**
This PR provides typings to be used in adapters for the smart home system [ioBroker](https://github.com/ioBroker/ioBroker/wiki/Adapter-Development-Documentation). Adapters (which work like installable plugins) have access to the core functionality through a file which is included in every adapter and links it to the core installation ([example](https://github.com/AlCalzone/ioBroker.tradfri/blob/master/src/lib/utils.ts)). I expose those in the global `ioBroker` namespace.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. **Not sure what to do here, this is not a change**
- [ ] Add or edit tests to reflect the change. **Also not sure what I'm supposed to test**
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package. **Note: Although the package `iobroker` exists on npm, it is only the installation routine and not a library. As such it is never `require()`d and does not need to have its own typings.**
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/iobroker/iobroker-tests.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+// Help! what to test here?
```
Comment (uniqueiniquity) [inline types/iobroker/iobroker-tests.ts]: I'm having trouble understanding a couple things:
- are these types describing the ioBroker package or the ioBroker.js-controller package? If the latter, the name should be changed
- How would you expect to use the types in this package? I can't quite tell from the docs how you would use it, since it seems like the ioBroker.js-controller package doesn't actually export anything. If it's meant to just be a collection of types available to someone who would be writing to match some sort of adapter spec, then you should just use as many of the types you are exporting in this test file so you know if future changes cause any regressions. 

Basically, in this file, you just want to use the package as a user would.

**Review by uniqueiniquity — CHANGES_REQUESTED** [review]

File: types/iobroker/iobroker-tests.ts (hunk 1 above)
Comment (AlCalzone) [reply types/iobroker/iobroker-tests.ts]: 1.:
An adapter is basically a plugin that is executed by the `iobroker.js-controller` in a separate process. Currently it using a file like [this one](https://github.com/AlCalzone/ioBroker.tradfri/blob/master/src/lib/utils.ts) to access certain methods/classes necessary for the execution, including the `Adapter` class I declared in these definitions.
As part of an upcoming refactoring, these files are no longer included in every adapter, but will in a separate package `iobroker.adapter-core` - I guess this should be the correct name for the typings then? I referenced `iobroker.js-controller` in the comments because it currently determines how the declarations have to look like.

2.:
An example usage can be seen in https://github.com/AlCalzone/ioBroker.tradfri/blob/master/src/main.ts  
The user imports `iobroker.adapter-core` and then creates an adapter instance:
```ts
import utils from "iobroker.adapter-core";

let adapter: utils.adapter({
	name: "tradfri",
	ready: () => { /* ... */ },
	// ...
});
```
The instance `adapter` is of the type `ioBroker.Adapter` declared in line 926, which is the most important part of this declaration. It has several event handlers and methods to interact with `js-controller`.

**Comment by AlCalzone** [comment]
@uniqueiniquity I added tests and changed some of the declarations. Hope this is ok now.

What about the package name? I still believe it should be called `iobroker`, because it acts as a framework for adapter devs like `@types/node`.

**Review by uniqueiniquity — APPROVED** [review]

**Comment by uniqueiniquity** [comment]
Thanks for clarifying. Looks great to me.

Congratulations on your first DefinitelyTyped contribution!
Thank you for being a part of the community!

**Comment by AlCalzone** [comment]
Thank you for the help!
